repo2docker/contentproviders/dataverse.py

```diff
@@ -0,0 +1,129 @@
+
+        # Check if the url matches any known Dataverse installation, bail if not.
+        host = next((host for host in self.hosts if url.startswith(host["url"])), None)
+        if host is None:
+            return
+
+        # Parse the url, to get the base for later API calls
+        parsed_url = urlparse(url)
+
+        # Corner case handling
+        if parsed_url.path.startswith("/file.xhtml"):
+            # There's no way of getting file information using its persistentId, the only thing we can do is assume that doi
+            # is structured as "doi:<dataset_doi>/<file_doi>" and try to handle dataset that way.
+            new_doi = doi.rsplit("/", 1)[0]
+            if new_doi == doi:
+                # tough luck :( Avoid inifite recursion and exit.
+                return
+            return self.detect(new_doi)
+        elif parsed_url.path.startswith("/api/access/datafile"):
+            # Raw url pointing to a datafile is a typical output from an External Tool integration
+            entity_id = os.path.basename(parsed_url.path)
+            search_query = "q=entityId:" + entity_id + "&type=file"
+            # Knowing the file identifier query search api to get parent dataset
+            search_url = urlunparse(
+                parsed_url._replace(path="/api/search", query=search_query)
+            )
+            resp = self.urlopen(search_url).read()
+            data = json.loads(resp.decode("utf-8"))["data"]
+            if data["count_in_response"] != 1:
+                self.log.debug("Dataverse search query failed!")
+                self.log.debug("  - doi = " + doi)
+                self.log.debug("  - url = " + url)
+                self.log.debug("  - resp = " + json.dumps(data))
+                return
+
+            self.record_id = deep_get(data, "items.0.dataset_persistent_id")
+        elif is_doi(doi):
+            self.record_id = "doi:" + normalize_doi(doi)
+
+        if hasattr(self, "record_id"):
+            return {"record": self.record_id, "host": host}
+
+    def fetch(self, spec, output_dir, yield_output=False):
+        """Fetch and unpack a Dataverse dataset."""
+        record_id = spec["record"]
+        host = spec["host"]
+
+        yield "Fetching Dataverse record {}.\n".format(record_id)
+        req = Request(
+            "{}/api/datasets/:persistentId?persistentId={}".format(
+                host["url"], record_id
+            ),
+            headers={"accept": "application/json"},
+        )
+        resp = self.urlopen(req)
+        record = json.loads(resp.read().decode("utf-8"))["data"]
+
+        # In order to fetch entire dataset we build a list of file IDs we want to fetch
+        # and then receive a zip file containing all of them.
+        # TODO: Dataverse has a limit for the zipfile size (see
+        # https://github.com/jupyter/repo2docker/pull/739#issuecomment-510834729)
+        # If size of the dataset is grater than 100MB individual files should be downloaded.
```

Comment:
Suggest to open an issue for that in repo2docker and remove the TODO here. 

OTOH, why not only implement an individual file download?

Reply:
I might be wrong here, but I got the impression that a scenario with a lot of small files is way more popular than just a few big files case in Dataverse repositories. Going with that hunch, I'd rather see one API call than a bazillion calls.

Reply:
Let me try to explain the pros and cons of downloading files individually from a Dataverse dataset vs. downloading them as a zip.

## Pros of downloading files individually

- Easier on Dataverse servers that store files on S3 because a "redirect" feature can be enabled such that files are downloading directly from S3 rather than passing through Dataverse servers.
- Easier on Dataverse servers because Dataverse servers do not have to create a zip file on the fly

## Cons of downloading files individually

- In order to know the file hierarchy, you have to parse metadata about the files and look for `directoryLabel` to get path/to/file information. (There can be other potentially interesting metadata about files in there, by the way, such as descriptions.)
- Lots of requests. Bazillion calls.

## Pros of downloading zip files

- Single request.
- Easier to deal with a single file. Just unpack it. The file hierarchy is in the zip file.

## Cons of downloading zip files

- May crash Dataverse servers. Dataverse servers have to construct the zip on the fly. Also when not using the S3 redirect option mentioned above, files pass through Glassfish and Apache.

I hope this helps! I might be forgetting a point or two.

Reply:
Let me try to explain the pros and cons of downloading files individually from a Dataverse dataset vs. downloading them as a zip.

## Pros of downloading files individually

- Easier on Dataverse servers that store files on S3 because a "redirect" feature can be enabled such that files are downloading directly from S3 rather than passing through Dataverse servers.
- Easier on Dataverse servers because Dataverse servers do not have to create a zip file on the fly

## Cons of downloading files individually

- In order to know the file hierarchy, you have to parse metadata about the files and look for `directoryLabel` to get path/to/file information. (There can be other potentially interesting metadata about files in there, by the way, such as descriptions.)
- Lots of requests. Bazillion calls.

## Pros of downloading zip files

- Single request.
- Easier to deal with a single file. Just unpack it. The file hierarchy is in the zip file.

## Cons of downloading zip files

- May crash Dataverse servers. Dataverse servers have to construct the zip on the fly. Also when not using the S3 redirect option mentioned above, files pass through Glassfish and Apache.

I hope this helps! I might be forgetting a point or two.

Reply:
Thanks for the elaborate description! That's very helpful.

Re. the `directoryLabel`: that is what is done now in the Zenodo and Figshare providers although I don't think directories are even possible in those repositories, see how `host["download"]` and `host["filename"]` are used in https://github.com/jupyter/repo2docker/blob/master/repo2docker/contentproviders/doi.py#L53

@betatim and @minrk should comment on what is preferable for BinderHub operators.

From a user perspective, I tend towards downloading files individually, because I want things not to break (e.g. if the ZIP would larger then 1 GB) and don't care about load on the servers. The comment says

> "Dataverse has a limit for the zipfile size"

so I think there is no way around implementing individual download (maybe as a second step if the zipfile download fails) unless you'd be fine with not supporting large repos.

Reply:
I mean, if it's easier to implement zip download first just to get something working, that's fine with me. It should work fine with datasets with small files. 😄 As you see, I'm just trying to explain the pitfalls of creating zip files on the fly from the Dataverse server perspective. (And it's a bad experience for users if the zip file isn't complete, if it only has 5 of the 17 files the user wants because they exceeded the zip size limit.) So yes, downloading individual files, even if it's a bazillion requests, is probably safer.

Reply:
I reimplemented fetch to use individual files download